src/constants/battler-tag-constants.ts

```diff
@@ -0,0 +1,84 @@
+// -- start tsdoc imports --
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import type { ElementalType } from "#enums/elemental-type";
+/* eslint-enable @typescript-eslint/no-unused-vars */
+// -- end tsdoc imports --
+
+import { BattlerTagType } from "#enums/battler-tag-type";
+
+/** All {@linkcode BattlerTagType | BattlerTagTypes} that grant semi-invulnerability */
+export const SEMI_INVULNERABLE_BATTLER_TAG_TYPES = Object.freeze([
+  BattlerTagType.FLYING,
+  BattlerTagType.UNDERGROUND,
+  BattlerTagType.UNDERWATER,
+  BattlerTagType.HIDDEN,
+]);
+
+/** All {@linkcode BattlerTagType | BattlerTagTypes} that grant protection/invulnerability */
+export const PROTECTION_BATTLER_TAG_TYPES = Object.freeze([
+  BattlerTagType.PROTECTED,
+  BattlerTagType.BANEFUL_BUNKER,
+  BattlerTagType.SPIKY_SHIELD,
+  BattlerTagType.SILK_TRAP,
+  BattlerTagType.BURNING_BULWARK,
+  BattlerTagType.KINGS_SHIELD,
+  BattlerTagType.OBSTRUCT,
+]);
+
+/** All {@linkcode BattlerTagType | BattlerTagTypes} that lock a move in.  */
+export const MOVE_LOCK_TAG_TYPES = Object.freeze([
+  BattlerTagType.FRENZY,
+  BattlerTagType.ROLLING,
+  BattlerTagType.UPROAR,
+  BattlerTagType.BIDE,
+]);
+
+/** All {@linkcode BattlerTagType | BattlerTagTypes} that boost critical hits. */
```

Comment:
```suggestion
/** All unstackable {@linkcode BattlerTagType | BattlerTagTypes} that boost critical hit rate. */
```
(G-max Chi Strike applies a `BattlerTagType.CRIT_BOOST_STACKABLE`, which is handled separately so that it can stack with itself.)